packages/cinterop/src/nativeDarwin/kotlin/io/realm/kotlin/internal/interop/RealmValueAllocator.kt

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright 2022 Realm Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+@file:Suppress("OVERRIDE_BY_INLINE")
+
+package io.realm.kotlin.internal.interop
+
+import kotlinx.cinterop.Arena
+import kotlinx.cinterop.addressOf
+import kotlinx.cinterop.alloc
+import kotlinx.cinterop.allocArray
+import kotlinx.cinterop.get
+import kotlinx.cinterop.pointed
+import kotlinx.cinterop.ptr
+import kotlinx.cinterop.set
+import kotlinx.cinterop.useContents
+import kotlinx.cinterop.usePinned
+import platform.posix.memcpy
+import realm_wrapper.realm_query_arg_t
+import realm_wrapper.realm_value_t
+import realm_wrapper.realm_value_type
+
+/**
+ * Only one allocator is needed for K/N as structs will be freed after completion no matter what,
+ * which includes their buffers.
+ */
+class NativeMemAllocator : MemTrackingAllocator {
+
+    val scope = Arena()
+
+    override inline fun allocRealmValueT(): RealmValueT = scope.alloc()
+
+    override fun nullTransport(): RealmValue =
+        createTransport(realm_value_type.RLM_TYPE_NULL)
+
+    override fun longTransport(value: Long): RealmValue =
+        createTransport(realm_value_type.RLM_TYPE_INT) {
+            integer = value
+        }
+
+    override fun booleanTransport(value: Boolean): RealmValue =
+        createTransport(realm_value_type.RLM_TYPE_BOOL) {
+            boolean = value
+        }
+
+    override fun timestampTransport(value: Timestamp): RealmValue =
+        createTransport(realm_value_type.RLM_TYPE_TIMESTAMP) {
+            timestamp.apply {
+                seconds = value.seconds
+                nanoseconds = value.nanoSeconds
+            }
+        }
+
+    override fun floatTransport(value: Float): RealmValue =
+        createTransport(realm_value_type.RLM_TYPE_FLOAT) {
+            fnum = value
+        }
+
+    override fun doubleTransport(value: Double): RealmValue =
+        createTransport(realm_value_type.RLM_TYPE_DOUBLE) {
+            dnum = value
+        }
+
+    override fun objectIdTransport(value: ByteArray): RealmValue =
+        createTransport(realm_value_type.RLM_TYPE_OBJECT_ID) {
+            object_id.apply {
```

Comment:
Can't we use the `memcpy` approach as below? (don't know if it is faster though 🙈 #TODOBENCHMARK 🤪 )